third_party/amd/lib/TritonAMDGPUToLLVM/TritonGPUToLLVMPass.cpp

```diff
@@ -983,6 +984,140 @@ struct ConvertTritonAMDGPUToLLVM
     return builder.create<triton::FpToFpOp>(loc, tensorPromotedType, operand);
   }
 
+  void promoteReduceOps(ModuleOp mod) const {
+
+    // promote reduce ops
+    mod.walk([&](triton::ReduceOp op) -> void {
+      OpBuilder builder(op);
+
+      // promote operands
+      SmallVector<Value> newOperands;
+      for (OpOperand &operand : op->getOpOperands()) {
+        auto val = operand.get();
+        auto oldType = val.getType().cast<RankedTensorType>();
+        auto elemType = oldType.getElementType();
+        if (elemType.isInteger(16) || elemType.isInteger(8)) {
+          auto newType = oldType.cloneWith(std::nullopt, builder.getIntegerType(32));
+          auto promotedVal =
+              builder.create<mlir::arith::ExtSIOp>(op->getLoc(), newType, val);
+          newOperands.push_back(promotedVal);
+        } else if (elemType.isF16()) {
+          auto newType = oldType.cloneWith(std::nullopt, builder.getF32Type());
+          auto promotedVal =
+              builder.create<mlir::arith::ExtFOp>(op->getLoc(), newType, val);
+          newOperands.push_back(promotedVal);
+        } else {
+          newOperands.push_back(val);
+        }
+      }
+      op->setOperands(newOperands);
+
+      // promote results
+      for (Value result : op.getResults()) {
+        auto type = result.getType();
+        if (type.isInteger(16)) {
+          result.setType(builder.getIntegerType(32));
+
+          // add trunc if float result type was changed
+          builder.setInsertionPointAfter(op);
```

Comment:
nit: can we merge this across all the if condition? It can even go before the loop

Reply:
on it